imap_processing/lo/l1b/lo_l1b.py

```diff
@@ -260,6 +266,127 @@ def set_spin_cycle(l1a_de: xr.Dataset, l1b_de: xr.Dataset) -> xr.Dataset:
     return l1b_de
 
 
+def get_spin_start_times(
+    l1a_de: xr.Dataset, l1b_de: xr.Dataset, spin_data: xr.Dataset, acq_end: xr.DataArray
+) -> xr.DataArray:
+    """
+    Get the start time for the spin that each direct event is in.
+
+    The resulting array of spin start times will be equal to the length of the direct
+    events. If two direct events occurred in the same spin, then there will be repeating
+    spin start times.
+
+    Parameters
+    ----------
+    l1a_de : xr.Dataset
+        The L1A DE dataset.
+    l1b_de : xr.Dataset
+        The L1B DE dataset.
+    spin_data : xr.Dataset
+        The L1A Spin dataset.
+    acq_end : xr.DataArray
+        The end acquisition times for each spin ASC.
+
+    Returns
+    -------
+    spin_start_time : xr.DataArray
+        The start time for the spin that each direct event is in.
+    """
+    met = l1a_de["met"].values
+    # Find the closest stop_acq for each shcoarse
+    closest_stop_acq_indices = np.abs(met[:, None] - acq_end.values).argmin(axis=1)
+    # There are 28 spins per epoch (1 aggregated science cycle)
+    # Set the spin_cycle_num to the spin number relative to the
+    # start of the ASC
+    spin_cycle_num = l1b_de["spin_cycle"] % 28
+    # Get the seconds portion of the start time for each spin
+    start_sec_spins = spin_data["start_sec_spin"].values[
+        closest_stop_acq_indices, spin_cycle_num
+    ]
+    # Get the subseconds portion of the spin start time and convert from
+    # microseconds to seconds
+    start_subsec_spins = (
+        spin_data["start_subsec_spin"].values[closest_stop_acq_indices, spin_cycle_num]
+        * 1e-6
+    )
+
+    # Combine the seconds and subseconds to get the start time for each spin
+    spin_start_time = start_sec_spins + start_subsec_spins
+    return xr.DataArray(spin_start_time)
+
+
+def set_event_met(
+    l1a_de: xr.Dataset,
+    l1b_de: xr.Dataset,
+    spin_start_time: xr.DataArray,
+    avg_spin_durations: xr.DataArray,
+) -> xr.Dataset:
+    """
+    Get the event MET for each direct event.
+
+    Each direct event is converted from a data number to engineering unit in seconds.
+    de_eu_time de_dn_time / 4096 * avg_spin_duration
```

Comment:
Great docstrings, makes it very clear exactly what is going on.

```suggestion
    de_eu_time = de_dn_time / 4096 * avg_spin_duration
```

Reply:
Thanks!